sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -278,23 +298,17 @@ object CTESubstitution extends Rule[LogicalPlan] {
         if cteRelations.exists(r => plan.conf.resolver(r._1, table)) =>
         throw QueryCompilationErrors.timeTravelUnsupportedError(toSQLId(table))
 
-      case u @ UnresolvedRelation(Seq(table), _, _) =>
-        cteRelations.find(r => plan.conf.resolver(r._1, table)).map { case (_, d) =>
-          if (alwaysInline) {
-            d.child
-          } else {
-            // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
-          }
-        }.getOrElse(u)
+      case u: UnresolvedRelation =>
+        resolveCTEInUnresolvedRelation(plan, alwaysInline, cteRelations, u)
 
       case p: PlanWithUnresolvedIdentifier =>
-        // We must look up CTE relations first when resolving `UnresolvedRelation`s,
-        // but we can't do it here as `PlanWithUnresolvedIdentifier` is a leaf node
-        // and may produce `UnresolvedRelation` later.
-        // Here we wrap it with `UnresolvedWithCTERelations` so that we can
-        // delay the CTE relations lookup after `PlanWithUnresolvedIdentifier` is resolved.
```

Comment:
This part of the comment needs an update, we now use a different `planBuilder` to delay the CTE lokup.